swconfig/files/switch.lua

```diff
@@ -0,0 +1,438 @@
+-- vim: set expandtab tabstop=2 shiftwidth=2 softtabstop=2 :
+rs232 = require("luars232")
+utils = require("swconfig.utils")
+
+local MAX_HOSTNAME_LEN              = 20
+local SERIAL_READ_ONE_CHAR_TIMEOUT  = 100
+local SERIAL_WRITE_ONE_CHAR_TIMEOUT = 100
+
+-- Switch class definition
+Switch = {}
+
+-- Yes, this is a way in lua to mimic an enum :'(
+-- It works because each value is a memory address that will be unique
+Switch.State = {
+  PRESS_ANY_KEY   = {},
+  USER_MAIN       = {},
+  ADMIN_MAIN      = {},
+  CONFIG          = {},
+  CONFIG_VLAN     = {},
+  CONFIG_IF       = {},
+  CONFIG_IF_RANGE = {},
+  LOGIN_USERNAME  = {},
+  LOGIN_PASSWORD  = {},
+  UNKNOWN         = {},
+}
+
+-- This method instantiates a new Switch object.
+-- It doesn't connect to the serial port, only prepares the object
+function Switch:new(file, username, password)
+  -- Following 3 lines are a way in lua to use OOP :'(
+  -- Note that we don't need inheritance feature here
+  local s = {}
+  setmetatable(s, self)
+  self.__index = self
+
+  s.baud = rs232.RS232_BAUD_115200
+  s.bits = rs232.RS232_DATA_8
+  s.parity = rs232.RS232_PARITY_NONE
+  s.flow = rs232.RS232_FLOW_OFF
+  s.stop_bits = rs232.RS232_STOP_1
+
+  s.file = file
+  s.sock = nil
+  s.hostname = nil
+  s.state = Switch.State.UNKNOWN
+  s.username = username
+  s.password = password
+  return s
+end
+
+-- This opens the serial connection
+function Switch:open()
+  if sock ~= nil then
+    return false
+  end
+
+  -- Ask the kernel to see if someone is already using the serial link
+  -- Better than a lock we would handle ourselves :)
+  if self:_file_is_busy() then
+    self:_print_error("'" .. self.file .. "' is busy")
+    return false
+  end
+
+  local err, sock = rs232.open(self.file)
+  if err ~= rs232.RS232_ERR_NOERROR then
+    self._print_error("Unable to open the serial connection to the switch", err)
+    return false
+  end
+
+  sock:set_baud_rate(self.baud)
+  sock:set_data_bits(self.bits)
+  sock:set_parity(self.parity)
+  sock:set_flow_control(self.flow)
+  sock:set_stop_bits(self.stop_bits)
+
+  self.sock = sock
+  return true
+end
+
+-- This closes the serial connection and resets the switch object for reuse
+-- It's very important NOT to logout from the switch
+-- If we logout 5 times, we get following warning:
+--  CLI is restarting too fast
+--  Console will be blocked 300 seconsd if you restart CLI less than 90 seconds next time.
+--
+-- If we logout a sixth time, UART CLI won't respond at all for 5 minutes
+--  Console is blocking 300 seconsd
+--
+-- It seems the best option is NOT to logout. CLI will auto logout after an idle time of 10min
+function Switch:close()
+  if self.sock ~= nil then
+    self.sock:close()
+    self.sock = nil
+  end
+
+  self.hostname = nil
+  self.state = Switch.State.UNKNOWN
+end
+
+-- Ask the kernel to see if someone else is already using the serial link
+function Switch:_file_is_busy()
+  fd = io.popen("lsof " .. self.file)
```

Comment:
Don't you need to check the value of fd ? I'm no lua expert but I can see people do things like `assert(io.popen("your_program"))`